paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt

```diff
@@ -637,7 +637,7 @@ class Paparazzi @JvmOverloads constructor(
       this
     }
 
-  companion object {
+  internal companion object {
```

Comment:
Keep this `public`, if someone added `fun Paparazzi.Companion.createMyPaparazzi()` to get `Paparazzi.createMyPaparazzi()` function, they'll break. Internals are hidden enough.

Reply:
cc @drewhamilton

Reply:
I don't follow, as @drewhamilton mentioned, all the members of this companion object are either private or internal.

Re: the internal members,
```
    internal val TIME_OFFSET_NANOS = TimeUnit.HOURS.toNanos(1L)
    internal lateinit var renderer: Renderer
    internal val isInitialized get() = ::renderer.isInitialized
    internal lateinit var sessionParamsBuilder: SessionParamsBuilder
```

...we certainly don't want these depended on or exposed.  What am I missing?

Reply:
Yeah, that is true, but you want to keep the class Paparazzi$Companion.class to exist as public, even though it's contents are (package) private. So people can attach extension functions on it. Without a public declaration there's no way. For example I think most java.* classes don't have companions, so it's impossible to create "static" extension functions, even if you wanted to. I'm afk, but I'm sure you can find some empty companions in other libraries.

Reply:
Try my extension fun out on the sample w/ and w/o companion.

Reply:
Discussed offline.

I hadn't realized how popular it is to use extension functions on companion objects as factory methods, i.e.,
```
fun Paparazzi.Companion.createPaparazzi() = Paparazzi(...)
```

I would have thought the unused receiver would make this a bit of an anti-pattern, but @TWiStErRob has convinced me that this pattern is quite helpful for namespacing reasons, vs using top-level functions which pollute the global namespace, i.e.,
```
fun createPaparazzi() = Paparazzi(...)
```

So I'm reverting this one!

Reply:
Not convinced that you need to do this.

Reply:
Can you elaborate more?  Not convinced that it should remain `internal` or revert back to `public`?

Reply:
Note: a few example from the "wild":

 * [Kotlin > Boolean](https://github.com/JetBrains/kotlin/blob/36d5ffd4eb646497f447495f3737cdf174eb3d1d/core/builtins/native/kotlin/Boolean.kt#L17)
 * [kotlinx-datetime > Clock](https://github.com/Kotlin/kotlinx-datetime/blob/aabed342cd1471b1138fc1130245c1a7a1c54a40/core/common/src/Clock.kt#L28)
 * [ktor > Url](https://github.com/ktorio/ktor/blob/dd9b42e3a548cd821fa8411801fcc1f8b83f61a1/ktor-http/common/src/io/ktor/http/Url.kt#L121)
 * [kotlin > TimeSource](https://github.com/kackey0-1/kotlin/blob/b5eafb9eb76311b9507786242a3f2b3b6021f0d0/libraries/stdlib/src/kotlin/time/TimeSource.kt#L104)
 * [mirai > MiraiLogger](https://github.com/killerprojecte/mirai/blob/e60384126b7e49fb00017d54dd6f80e332d2d927/mirai-core-api/src/commonMain/kotlin/utils/MiraiLogger.kt#L69)

Reply:
I suppose the list of counter examples would then be every other class written ever? So I think that wins by many orders of magnitude.

Hanging an empty companion is not your problem. It's [the language's problem](https://github.com/Kotlin/KEEP/issues/348).

Reply:
Gotcha.  Makes sense.  Reapplying `internal`.